### PR TITLE
fix: provider connect method accept options

### LINKED ISSUE #1396: Accept `ProviderOptions` via `Provider.connect`
Currently, it is possible to pass in some `ProviderOptions` for customization via `Provider.create`:
```ts
const provider = await Provider.create(url, { fetch: customFetch });
```
but not via `Provider.connect`
```ts
await provider.connect(newNetworkUrl);
```

This should be possible.

### PR DESCRIPTION
**Pull request by LuizAsFight**
I can create a `Provider` passing custom client options with 
```
const provider = await Provider.create(url, { fetch: customFetch });
```

but when connecting to a new network url, we cannot pass the options anymore
```
await provider.connect(newNetworkUrl);
```

---------------
this PR aims to
- allow `connect` method to receive `options`

Closes #1396

### CODE REVIEW

File: packages/providers/src/provider.ts [hunk 1]
```diff
@@ -310,9 +310,9 @@ export default class Provider {
   /**
    * Updates the URL for the provider and fetches the consensus parameters for the new URL, if needed.
    */
-  async connect(url: string) {
+  async connect(url: string, options?: ProviderOptions) {
     this.url = url;
-    this.operations = this.createOperations(url);
+    this.operations = this.createOperations(url, options);
     await this.fetchChainAndNodeInfo();
   }
```
Comment (luizstacio) [inline packages/providers/src/provider.ts:317]: Should we save the options when we create the instance and in this case just forward the options?

**Review by luizstacio — CHANGES_REQUESTED** [review]

File: packages/providers/src/provider.ts (hunk 1 above)
Comment (LuizAsFight) [reply packages/providers/src/provider.ts:317]: I think it doesn't make sense as different networks can have different fetch options

Comment (luizstacio) [reply packages/providers/src/provider.ts:317]: In this case, you would create a completely new provider... we can also make it optional and use `options || this.options`.

**Comment by Dhaiwat10** [comment]
Thanks @LuizAsFight, I'll get this one over the line

File: packages/providers/src/provider.ts [hunk 2]
```diff
@@ -243,7 +243,7 @@ export default class Provider {
    * @param options - Additional options for the provider
    * @hidden
    */
-  protected constructor(
```
Comment (Dhaiwat10) [inline packages/providers/src/provider.ts:246]: @LuizAsFight I'm curious - why did you remove the `protected` modifier from the constructor here?

File: packages/providers/src/provider.ts (hunk 1 above)
Comment (Dhaiwat10) [reply packages/providers/src/provider.ts:317]: I think saving the options makes sense

File: packages/providers/src/provider.ts (hunk 2 above)
Comment (arboleya) [reply packages/providers/src/provider.ts:246]: Yes, I'm assuming it was unintentional.

That should be kept.

File: packages/providers/src/provider.ts (hunk 1 above)
Comment (arboleya) [reply packages/providers/src/provider.ts]: Yep, and also giving the option to override it on the `connect` method:

```ts
this.operations = this.createOperations(url, options ?? this.options);
```

File: packages/providers/src/provider.ts (hunk 2 above)
Comment (LuizAsFight) [reply packages/providers/src/provider.ts:246]: yep can remove it @Dhaiwat10

**Review by arboleya — APPROVED** [review]

**Review by Torres-ssf — APPROVED** [review]

**Pull request merged**